### DISCUSSION TITLE
What is the best practice for zero downtime deployment of a Docker service?

### DISCUSSION THREAD

**Discussion by callmewhy** [Dokploy/dokploy#492] [Q&A] [answered]
In the past, I used nginx + docker compose for deployment, and there would be a few seconds of service unavailability with each deployment. 

I'd like to ask what is the best practice for zero downtime in dokploy? Is it traefik + replica deployment? 

If it could be updated in the documentation like [dokku](https://dokku.com/docs/deployment/zero-downtime-deploys/), that would b better. 

Thanks

**Comment by Siumauricio** [comment]
Hey, We don't have docs for zero downtime yet, you can achieve this same behavior playing with Swarm Settings, I will try this week to create some example in the docs


![image](https://github.com/user-attachments/assets/19870c42-0dcb-4091-a835-c28f6b0fdb65)

**Reply by johngrimsey** [reply]
I would also love to read these docs @Siumauricio! 🙏

**Comment by Siumauricio** [comment, marked as answer]
I leave you with this simple to understand example, 

we will use a basic express app, which does 2 things

https://github.com/Dokploy/swarm-test

1. it has 1 endpoint called `/health` [Endpoint](https://github.com/Dokploy/swarm-test/blob/main/index.js#L20) which is the one that will tell us if our application is healthy.
2. for testing purposes I added a delay at the start of the application to simulate when our application is restarting or deploying a new version  [Delay](https://github.com/Dokploy/swarm-test/blob/main/index.js#L9)

We are going to use Nixpacks as builder and your config should similar to this 

![image](https://github.com/user-attachments/assets/c3ba98ef-829d-4da2-8477-c292854ca284)


If you want to test that there is no zero downtime yet, you can simply deploy the application and then create another deployment and while doing the deployment reload the page in the path /health and you will see that a bad gateway will appear.

Now go to the advanced section of our application, and go to the Swarm Settings section, we are going to modify the first section of Healtchecks.

We will use this configuration specifically, paste and save it
```js
{
  “Test": [
    “CMD”,
    “curl”,
    “-f”,
    “http://localhost:3000/health”
  ],
  “Interval": 30000000000,
  “Timeout": 10000000000,
  “StartPeriod": 30000000000,
  “Retries": 3
}
```

This configuration basically tells to Docker to do:

1.  Make a request inside the container to `http://localhost:3000/health` and then we are also saying to make in interval of 30000000000 nanosec, and also makes 3 retries before switching to the new container

![image](https://github.com/user-attachments/assets/4112effc-6413-40b2-894f-891e6e7869cc)

that would be all, now to test it just do the same as I said above deploy your application 2 times and you will see that you will never get a bad gateway while the deployment is happening.

This is something basic, however you can improve it using other configurations like rollback configs, restart policy and many other things.

**Reply by johngrimsey** [reply]
🙏

**Reply by ardasevinc** [reply]
~~can we modify this setting without setting a registry?~~

yes you can, just tested.

**Reply by HanifCarroll** [reply]
Where is this setting? In the advanced tab I only see Run Command, Volumes, Import, and Enable Isolated Deployment. Is it not available via docker compose deployments and therefore there's no way to do zero downtime deployments with compose?
